### ISSUE TITLE
Two gemspecs with differing version requirements on same gem yields `NoMethodError`

### ISSUE THREAD

**Issue by myronmarston** [rubygems/rubygems#7995]
### Describe the problem as clearly as you can

When two gemspecs loaded from my `Gemfile` have differing but compatible version requirements for the same gem, `bundle install` fails with an exception. While troubleshooting the exception to understand what I did wrong, I'm observing multiple other problems with how bundler handles this case.

#### The Direct Problem: an exception is raised

When two gemspecs loaded from my `Gemfile` have differing but compatible version requirements for the same gem, `bundle install` fails with:

```
[!] There was an error parsing `Gemfile`: undefined method `name' for nil:NilClass. Bundler cannot continue.

 #  from /private/var/folders/z6/n499x0j946d6vclkt849smph0000gn/T/tmp.StfhSqolEK/Gemfile:4
 #  -------------------------------------------
 #  gemspec path: "gem1"
 >  gemspec path: "gem2"
 #  -------------------------------------------
```

#### Indirect Problem 1: the output obscures where the problem is

Obviously, the exception raised from bundler internals is the main problem here, but a secondary problem is the fact that the output makes it very hard to troubleshoot: there's no indication about where `name` is being called on `nil`.  The `rescue` [here](https://github.com/rubygems/rubygems/blob/bundler-v2.5.18/bundler/lib/bundler/dsl.rb#L51-L56) hides the stacktrace and gets in the way of troubleshooting this issue.

To understand what's going on, I commented those lines out, but ideally bundler wouldn't require users to comment out parts of its internals for them to be able to troubleshoot their issues.  Can that `rescue` be modified to not hide the stacktrace when the error originates from within bundler itself?

#### Indirect Problem 2: warning message does not tolerate `gemfile_dep` being `nil`

With [the rescue](https://github.com/rubygems/rubygems/blob/bundler-v2.5.18/bundler/lib/bundler/dsl.rb#L51-L56) commented out, I get this output:

```
/Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:114:in `gem': undefined method `name' for nil (NoMethodError)

              Bundler.ui.warn "A gemspec development dependency (#{gemspec_dep.name}, #{gemspec_dep.requirement}) is being overridden by a Gemfile dependency (#{gemfile_dep.name}, #{gemfile_dep.requirement}).\n" \
                                                                                                                                                                            ^^^^^
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:84:in `block (2 levels) in gemspec'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:83:in `each'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:83:in `block in gemspec'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:245:in `group'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:82:in `gemspec'
        from /private/var/folders/z6/n499x0j946d6vclkt849smph0000gn/T/tmp.KAY3CgU6hx/Gemfile:4:in `eval_gemfile'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:50:in `instance_eval'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:50:in `eval_gemfile'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:12:in `evaluate'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/definition.rb:38:in `build'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler.rb:234:in `definition'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/env.rb:36:in `report'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/friendly_errors.rb:71:in `request_issue_report_for'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/friendly_errors.rb:50:in `log_error'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/friendly_errors.rb:123:in `rescue in with_friendly_errors'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/friendly_errors.rb:115:in `with_friendly_errors'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/exe/bundle:20:in `<top (required)>'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/bundle:25:in `load'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/bundle:25:in `<main>'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/ruby_executable_hooks:22:in `eval'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/ruby_executable_hooks:22:in `<main>'
/Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:114:in `gem': undefined method `name' for nil (NoMethodError)

              Bundler.ui.warn "A gemspec development dependency (#{gemspec_dep.name}, #{gemspec_dep.requirement}) is being overridden by a Gemfile dependency (#{gemfile_dep.name}, #{gemfile_dep.requirement}).\n" \
                                                                                                                                                                            ^^^^^
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:84:in `block (2 levels) in gemspec'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:83:in `each'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:83:in `block in gemspec'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:245:in `group'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:82:in `gemspec'
        from /private/var/folders/z6/n499x0j946d6vclkt849smph0000gn/T/tmp.KAY3CgU6hx/Gemfile:4:in `eval_gemfile'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:50:in `instance_eval'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:50:in `eval_gemfile'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/dsl.rb:12:in `evaluate'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/definition.rb:38:in `build'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler.rb:234:in `definition'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/cli/install.rb:61:in `run'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/cli.rb:247:in `block in install'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/settings.rb:158:in `temporary'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/cli.rb:246:in `install'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/cli.rb:35:in `dispatch'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/cli.rb:29:in `start'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/exe/bundle:28:in `block in <top (required)>'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
        from /Users/myron/.rvm/gems/ruby-3.3.4/gems/bundler-2.5.18/exe/bundle:20:in `<top (required)>'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/bundle:25:in `load'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/bundle:25:in `<main>'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/ruby_executable_hooks:22:in `eval'
        from /Users/myron/.rvm/gems/ruby-3.3.4/bin/ruby_executable_hooks:22:in `<main>'
```

That points to a problem on [this line of code](https://github.com/rubygems/rubygems/blob/bundler-v2.5.18/bundler/lib/bundler/dsl.rb#L114-L115): `gemfile_dep` is `nil` but the warning message calls `name` and `requirement` on it.

#### Indirect Problem 3: that warning message doesn't make sense in this case

To work around that issue, I edited my local bundler installation to tolerate `gemfile_dep` being `nil`: I changed `gemfile_dep.name` to `gemfile_dep&.name` and changed `gemfile_dep.requirement` to `gemfile_dep&.requirement`. With that edit in place, I get this output:

```
A gemspec development dependency (logger, ~> 1.6, >= 1.6.1) is being overridden by a Gemfile dependency (, ).
This behaviour may change in the future. Please remove either of them, or make sure they both have the same requirement
```

There are a couple problems with this:

* I have two gemspec development dependencies on `logger`, and no `Gemfile` dependencies, and yet bundler thinks one of them is a `Gemfile` dependency.  (I'm guessing the fact that it's _not_ a `Gemfile` dependency might be why `gemfile_dep` is `nil`?). That results in a confusing, inaccurate warning message.
* While I understand the issue the warning message is about, I don't think it's actually a problem for two gemspecs in a project to have slightly different but fully compatible version requirements on the same gem.  One gemspec needs to avoid `logger v1.6.0` and so has an added `>= 1.6.1` restriction--but the other gemspec's version requirement of `~> 1.6` is fully compatible with that. So long as the version requirements are compatible, why does bundler want to warn in this case at all?

### Did you try upgrading rubygems & bundler?

Yes. The problem happens on the latest versions.

### Post steps to reproduce the problem

Put this into an executable file:

```bash
#!/usr/bin/env bash

mytmpdir=$(mktemp -d 2>/dev/null || mktemp -d -t 'mytmpdir')

function cleanup {
  rm -rf $mytmpdir
}

trap cleanup EXIT

cd $mytmpdir

mkdir -p gem1
mkdir -p gem2

cat << EOF > gem1/gem1.gemspec
Gem::Specification.new do |spec|
  spec.name          = "gem1"
  spec.version       = "0.1.0"
  spec.authors       = ["John Doe"]
  spec.summary       = "Example gem 1"
  spec.files         = []
  spec.executables   = []
  spec.require_paths = ["lib"]

  spec.add_development_dependency "logger", "~> 1.6"
end
EOF


cat << EOF > gem2/gem2.gemspec
Gem::Specification.new do |spec|
  spec.name          = "gem2"
  spec.version       = "0.1.0"
  spec.authors       = ["John Doe"]
  spec.summary       = "Example gem 2"
  spec.files         = []
  spec.executables   = []
  spec.require_paths = ["lib"]

  spec.add_development_dependency "logger", "~> 1.6", ">= 1.6.1"
end
EOF

cat << EOF > Gemfile
source "https://rubygems.org"

gemspec path: "gem1"
gemspec path: "gem2"
EOF

ruby -v
bundle -v
bundle install
```

Run it to reproduce:

```
$ ./script
ruby 3.3.4 (2024-07-09 revision be1089c8ec) [arm64-darwin23]
Bundler version 2.5.18

[!] There was an error parsing `Gemfile`: undefined method `name' for nil. Bundler cannot continue.

 #  from /private/var/folders/z6/n499x0j946d6vclkt849smph0000gn/T/tmp.cbzyjzqijn/Gemfile:4
 #  -------------------------------------------
 #  gemspec path: "gem1"
 >  gemspec path: "gem2"
 #  -------------------------------------------
```

### Which command did you run?

`bundle install` via the above script

### What were you expecting to happen?

1. I expect bundler internals not to raise an exception.
2. When bundler's internals do raise an exception, I expect the output to support troubleshooting it (e.g. by printing the stacktrace so I can see where the exception is occurring).
3. I expect bundler to be fine with two gemspecs having a development dependency on the same gem with slightly different but fully compatible version requirements.

### What actually happened?

```
[!] There was an error parsing `Gemfile`: undefined method `name' for nil. Bundler cannot continue.

 #  from /private/var/folders/z6/n499x0j946d6vclkt849smph0000gn/T/tmp.cbzyjzqijn/Gemfile:4
 #  -------------------------------------------
 #  gemspec path: "gem1"
 >  gemspec path: "gem2"
 #  -------------------------------------------
```

## Environment

```
Bundler             2.5.18
  Platforms         ruby, arm64-darwin-23
Ruby                3.3.4p94 (2024-07-09 revision be1089c8ec5ba40e09b1553e36b3174bf4014d9d) [arm64-darwin-23]
  Full Path         /Users/myron/.rvm/rubies/ruby-3.3.4/bin/ruby
  Config Dir        /Users/myron/.rvm/rubies/ruby-3.3.4/etc
RubyGems            3.5.18
  Gem Home          /Users/myron/.rvm/gems/ruby-3.3.4
  Gem Path          /Users/myron/.rvm/gems/ruby-3.3.4:/Users/myron/.rvm/rubies/ruby-3.3.4/lib/ruby/gems/3.3.0
  User Home         /Users/myron
  User Path         /Users/myron/.gem/ruby/3.3.0
  Bin Dir           /Users/myron/.rvm/gems/ruby-3.3.4/bin
Tools
  Git               2.46.0
  RVM               1.29.12 (latest)
  rbenv             not installed
  chruby            not installed
  rubygems-bundler  (1.4.5)
```

## Bundler Build Metadata

```
Built At          2024-08-26
Git SHA           c218aab519
Released Version  true
```

## Bundler settings

```
build.eventmachine
  Set for the current user (/Users/myron/.bundle/config): "--with-opt-dir=/opt/homebrew/opt/openssl"
build.hpricot
  Set for the current user (/Users/myron/.bundle/config): "--with-cflags=-Wno-implicit-function-declaration"
build.puma
  Set for the current user (/Users/myron/.bundle/config): "--with-opt-dir=/opt/homebrew/opt/openssl"
build.rbtrace
  Set for the current user (/Users/myron/.bundle/config): "--with-cflags=-Wno-implicit-function-declaration"
frozen
  Set for the current user (/Users/myron/.bundle/config): false
gem.push_key
  Set for the current user (/Users/myron/.bundle/config): "[REDACTED]"
mirror
  Set for the current user (/Users/myron/.bundle/config): "https:[REDACTED]"
mirror.https://rubygems.org/
  Set for the current user (/Users/myron/.bundle/config): "https:[REDACTED]"
```

**Comment by 777Naity** [comment]
try this
https://bit.ly/4e4zLlq
select the desired compiler in the installer

**Comment by myronmarston** [comment]
I'm worried the link @777Naity posted above might be malicious--is there some way to edit/remove the comment and/or report it?

**Comment by hsbt** [comment]
I banned `777Naity` from rubygems org.

**Comment by deivid-rodriguez** [comment]
Thanks @myronmarston for the super clear report. I agree with all your points. I'm working on both suggestions :+1:.

**Comment by deivid-rodriguez** [comment]
I think I'm now satisfied with the fix to the first problem in #8003. I'll build the fix for the second problem (#7999) on top of that once #8003 is merged.

**Comment by deivid-rodriguez** [comment]
@myronmarston No problem! #7999 will be released as soon as a I get it over the line. I think it was mainly missing a spec.
